feat(metrics): Add team_key_transaction fake mri transformation

iambriccardo opened the pull request:
This PR aims at solving the issue in which the `team_key_transaction` derived operation doesn't require a `metric_mri` but because of the `MetricField` it actually requires it but without using it. This doesn't create problems until the `team_key_transaction` is used in the `orderby`. In that case because the `orderby` validation (available [here](https://github.com/getsentry/sentry/blob/fb2a714db17be29d29d5ae3735651dff3b9066b0/src/sentry/snuba/metrics/query.py#L240-L240)) requires all the fields in the `orderby` to have the same entity we cannot always pass the dummy `TransactionMRI.DURATION` to the `team_key_transaction` because if we are going to use for example the `TransactionMRI.USER_MISERY` inside of the `orderby` we are going to work with a `set` entity instead of a `distribution` entity.

There were several solutions to this problem which included the rework of some metrics layer internals, however with @ahmedetefy we came to the conclusion that we can perform an automatic dummy mri injection in the transformer instead of changing the core code. This means that whenever `team_key_transaction` is used and fed into the transformer, the following special mri should be used `e:transactions/team_key_transaction@none`. Under the hood the transformer will infer the proper mri to add to each instance of `team_key_transaction` based on the entity type inferred from the `select` fields.

Commit:
feat(metrics): Add team_key_transaction fake mri transformation

ahmedetefy commented on src/sentry/snuba/metrics/mqb_query_transformer.py:
```diff
@@ -9,17 +9,23 @@
 from sentry.snuba.metrics.query import OrderBy as MetricOrderBy
 from sentry.snuba.metrics.query_builder import FUNCTION_ALLOWLIST
 
+TEAM_KEY_TRANSACTION_FAKE_MRI = "e:custom/team_key_transaction@reserved"
```
```suggestion
TEAM_KEY_TRANSACTION_FAKE_MRI = "e:transactions/team_key_transaction@none"
```

ahmedetefy commented on src/sentry/snuba/metrics/mqb_query_transformer.py:
```diff
@@ -261,4 +422,17 @@ def transform_mqb_query_to_metrics_query(query: Query) -> MetricsQuery:
+        mq_dict["groupby"],
+        mq_dict["orderby"],
+    )
+    mq_dict["select"] = select
+    mq_dict["where"] = where
+    mq_dict["groupby"] = groupby
+    mq_dict["orderby"] = orderby
```
```suggestion
    mq_dict.update({"select": select, "where": where, "groupby": groupby, "orderby":orderby})
```

wmak committed:
fix(mep): Add user misery sort test

- This adds a test that sorts by both team_key_transactions and
  user_misery since the metrics layer doesn't currently support that

ahmedetefy commented on src/sentry/snuba/metrics/mqb_query_transformer.py:
```diff
@@ -235,6 +241,161 @@ def _transform_orderby(query_orderby):
+    return None
+
+
+def _derive_mri_to_apply(select, orderby):
+    mri_to_apply = TransactionMRI.DURATION.value
+
+    # We first check if there is an order by field that has the team_key_transaction, otherwise
+    # we just use the default mri of duration.
+    has_order_by_tkt = False
+    if orderby is not None:
+        for orderby_field in orderby:
+            if orderby_field.field.op == TEAM_KEY_TRANSACTION_OP:
+                has_order_by_tkt = True
+                break
+
+    if has_order_by_tkt:
+        mri_types = dict()
+        for select_field in select:
+            if select_field.op != TEAM_KEY_TRANSACTION_OP:
+                # We assume to have a correct MRI.
+                mri_type = select_field.metric_mri.split(":")[0]
+                if mri_type not in mri_types:
+                    # We set the mri to be the first occurrence, in order to make it easier for the user.
+                    #
+                    # It is important to note that in case of derived metrics we are going to recursively obtain the
+                    # leftmost ingested mri available in the dependency tree.
+                    mri_types[mri_type] = (
+                        select_field.metric_mri
+                        if mri_type != "e"
+                        else _recursively_compute_ingested_mri(select_field.metric_mri)
+                    )
+
+        if len(mri_types) == 1:
+            # In order to simplify the code we just set the MRI of the team_key_transaction to be equal
+            # to the one of the entity in the select in order to make sure that the "get_entity" method
+            # will always return the same result.
+            mri_to_apply = mri_types[list(mri_types.keys())[0]]
+
+    return mri_to_apply
```
Lets please replace this with the call `.get_entity()`

ahmedetefy commented on src/sentry/snuba/metrics/mqb_query_transformer.py:
```diff
@@ -235,6 +241,161 @@ def _transform_orderby(query_orderby):
+    return mri_to_apply
+
+
+def _tmp_alias_to_none(alias):
```
Lets just set a default alias to `team_key_transaction`

Commit:
Merge branch 'wmak/fix/add-user-misery-sort-test' into riccardo/fix-user-misery

Commit:
Fix implementation

Commit:
Improve code

Commit:
Add data to the mri

wmak commented on tests/snuba/api/endpoints/test_organization_events_mep.py:
```diff
@@ -442,6 +442,105 @@ def test_performance_homepage_query(self):
             assert field_meta["failure_count()"] == "integer"
 
+    def test_user_misery_and_team_key_sort(self):
+        self.features["organizations:use-metrics-layer"] = True
```
Can you remove this? The `OrganizationEventsMetricsEnhancedPerformanceEndpointTestWithMetricLayer` class already adds this automatically

iambriccardo replied:
Sure! Sorry for that

wmak reviewed (approved):
1 minor comment otherwise changes to the Visibility code looks good to me

Commit:
Change data structure

Commit:
Remove flag

Commit:
Merge branch 'master' into riccardo/fix-user-misery

andriisoldatenko commented on src/sentry/snuba/metrics/mqb_query_transformer.py:
```diff
@@ -235,6 +240,139 @@ def _transform_orderby(query_orderby):
+
+    # We first check if there is an order by field that has the team_key_transaction, otherwise
+    # we just use the default mri of duration.
+    has_order_by_tkt = False
```
```suggestion
    has_order_by_team_key_tx = False
```

andriisoldatenko replied:
just to be less cryptic.

andriisoldatenko commented on src/sentry/snuba/metrics/mqb_query_transformer.py:
```diff
@@ -235,6 +240,139 @@ def _transform_orderby(query_orderby):
+    if select is None:
+        return select
+
+    def _select(select_field):
```
```suggestion
    def _select_func(select_field):
```
it's a good idea to name it with suffix, especially when you use in map call later.

andriisoldatenko commented on src/sentry/snuba/metrics/mqb_query_transformer.py:
```diff
@@ -235,6 +240,139 @@ def _transform_orderby(query_orderby):
+
+        return select_field
+
+    return list(map(_select, select))
```
```suggestion
    return list(map(_select_func, select))
```

andriisoldatenko commented on src/sentry/snuba/metrics/mqb_query_transformer.py:
```diff
@@ -235,6 +240,139 @@ def _transform_orderby(query_orderby):
+    if where is None:
+        return where
+
+    def _where(where_field):
```
```suggestion
    def _where_func(where_field):
```
or another func-style idea came to my mind:
```
_where_predicate
```
but it's up to you.

Commit:
Fix PR comments

Commit:
Handle edge case

Commit:
Merge branch 'master' into riccardo/fix-user-misery
